python/src/scipp/compat/mantid.py

```diff
@@ -193,17 +193,119 @@ def validate_and_get_unit(unit):
         return known_units[unit]
 
 
-def init_pos(ws):
-    nHist = ws.getNumberHistograms()
-    pos = np.zeros([nHist, 3])
+def _to_array(positions, idx, pos, spec_idx):
+    positions[idx, :] = np.array([spec_idx, pos.X(), pos.Y(), pos.Z()])
+
+
+def _to_spherical(input):
+    transformed = np.ones(shape=input.shape)
+    transformed[:, 0] = input[:, 0]  # copy metadata
+    transformed[:, 1] = np.sqrt(np.sum(input[:, 1:]**2, axis=1))  # r
+    transformed[:, 2] = np.arccos(input[:, 3] /
+                                  transformed[:, 1])  # arccos ( z / r)
+    transformed[:, 3] = np.arctan2(input[:, 2], input[:, 1])  # arctan2(y, x)
+    return transformed
 
+
+def rotation_matrix_from_vectors(vec1, vec2):
+    """ Find the rotation matrix that aligns vec1 to vec2
+    :param vec1: A 3d "source" vector
+    :param vec2: A 3d "destination" vector
+    :return mat: Tranform (3x3) which aligns it with vec2.
+    """
+    a, b = (vec1 /
+            np.linalg.norm(vec1)).reshape(3), (vec2 /
+                                               np.linalg.norm(vec2)).reshape(3)
+    v = np.cross(a, b)
+    c = np.dot(a, b)
+    s = np.linalg.norm(v)
+    if s == 0:
+        return np.eye(3)
+    kmat = np.array([[0, -v[2], v[1]], [v[2], 0, -v[0]], [-v[1], v[0], 0]])
+    rotation_matrix = np.eye(3) + kmat + kmat.dot(kmat) * ((1 - c) / (s**2))
+    return rotation_matrix
+
+
+def init_pos(ws, source_pos, sample_pos):
+    from scipp import Dim
+    import numpy as np
     spec_info = ws.spectrumInfo()
-    for i in range(nHist):
-        if spec_info.hasDetectors(i):
-            p = spec_info.position(i)
-            pos[i, :] = [p.X(), p.Y(), p.Z()]
+    det_info = ws.detectorInfo()
+    total_detectors = spec_info.detectorCount()
+    non_detectors = sum(not spec.hasDetectors for spec in spec_info)
+    det_pos = np.zeros([total_detectors, 3 + 1])
+    no_detector_spectrum = np.full((non_detectors, 3 + 1), np.nan)
```

Comment:
Why is this needed?

Reply:
There are some spectra that do not have any detectors. We ultimately still need to provide a position for those spectra.

Reply:
It should be `nan`, no? `groupby.mean` should just do this automatically. See last example in https://scipp.github.io/user-guide/groupby.html.